pod-operation/src/components/pressure_transducer.rs

```diff
@@ -2,51 +2,97 @@ use ina219::INA219;
 use rppal::i2c::I2c;
 use tracing::debug;
 
-pub struct PressureTransducer {
-	ina: INA219<I2c>,
-}
-
 // The calibration value is used to adjust the maximum current measurement
 // and precision of measurements.
 const INA219_CALIBRATION_VALUE: u16 = 0xffff;
 
+// The pod will be using two INA219s, so we'll need to differentiate them
+// via their device address.
+const INA219_UPSTREAM_ADDRESS: u8 = 0x40;
+const INA219_DOWNSTREAM_ADDRESS: u8 = 0x41;
+
 // Even with the calibration values, the readings from the INA219 are not in
 // mA. A scaling factor is needed in order to convert the raw reading to mA and
 // this is not provided in the INA219 library that we are using. Note that this
 // value changes according to the calibration value. The exact formula can be
 // found in the INA219 datasheet.
 const INA219_SCALING_VALUE: f32 = 160.0;
 
-// The pressure transducer outputs a current between 4 mA and 20 mA with 0 PSI
-// and 300 PSI respectively. Assuming a linear interpolation, a 1 mA increase
-// results in a 18.75 PSI increase.
-const REF_CURRENT_LOW: f32 = 4.0;
-const REF_CURRENT_HIGH: f32 = 20.0;
-const REF_PRESSURE_LOW: f32 = 0.0;
-const REF_PRESSURE_HIGH: f32 = 300.0;
+struct Reference {
+	pressure_lo: f32,
+	pressure_span: f32,
+	current_lo: f32,
+	current_span: f32,
+}
 
-const REF_CURRENT_SPAN: f32 = REF_CURRENT_HIGH - REF_CURRENT_LOW;
-const REF_PRESSURE_SPAN: f32 = REF_PRESSURE_HIGH - REF_PRESSURE_LOW;
+impl Reference {
+	fn new(pressure_lo: f32, pressure_hi: f32, current_lo: f32, current_hi: f32) -> Self {
+		Self {
+			pressure_lo,
+			current_lo,
+			pressure_span: pressure_hi - pressure_lo,
+			current_span: current_hi - current_lo,
+		}
+	}
+}
 
-impl PressureTransducer {
-	pub fn new(ina219_addr: u8) -> Self {
-		let device = I2c::new().unwrap();
+fn init_ina(device_address: u8) -> INA219<I2c> {
+	let device = I2c::new().unwrap();
+
+	let mut ina219 = INA219::new(device, device_address);
+	debug!("Initialized I2C and INA219");
 
-		let mut ina219 = INA219::new(device, ina219_addr);
-		debug!("Initialized I2C and INA219");
+	ina219.calibrate(INA219_CALIBRATION_VALUE).unwrap();
+	debug!("Calibrating INA219");
 
-		ina219.calibrate(INA219_CALIBRATION_VALUE).unwrap();
-		debug!("Calibrating INA219");
+	ina219
+}
+
+pub struct PressureTransducer {
+	ina: INA219<I2c>,
+	ref_values: Reference,
+}
 
-		PressureTransducer { ina: ina219 }
+impl PressureTransducer {
+	// This constructor should be used for INA219s where the address pins are
+	// grounded. That is, the device address is 0x40.
+	pub fn upstream() -> Self {
+		// The upstream pressure transducer outputs a current between 4 mA and 20 mA
+		// with 0 PSI and 300 PSI respectively.
+		let upstream_ref = Reference::new(0.0, 300.0, 4.0, 20.0);
```

Comment:
Suggestion: could we make the reference values top-level constants so we don't have these magic values in the middle of the file?

Reply:
Unfortunately, I couldn't. I tried putting it at the top level as it was before, but the compiler complained, saying that the `Reference` constructor had to be `const`. When I made it `const` though, another error appeared saying the floating-point arithmetic in the constructor isn't allowed.

I suppose I could make the constants `i32` since they are already integers, but will there ever be a case in where the values won't be integers?